src/docs/sdk/performance/dynamic-sampling-context.mdx

```diff
@@ -0,0 +1,158 @@
+
+</Alert>
+
+Until now, traces sampling was only done through a `traces_sample_rate` option in the SDKs.
+This has quite a few drawbacks for users of Sentry SDKs:
+
+- Changing the sampling rate involved either redeploying applications (which is problematic in case of applications that are not updated automatically, i.e., mobile apps or physically distributed software) or building complex systems to dynamically fetch a sampling rate.
+- Sampling only happened based on a factor of randomness.
+  Employing sampling rules, for example, based on event parameters, is currently very complex.
+  While writing rules for singular **transactions** is possible, enforcing them on an entire **trace** is infeasable.
+
+The solution for these problems is **Dynamic Sampling**.
+Dynamic Sampling allows users to configure **sampling rules** directly in the Sentry interface. Important: Sampling rules may also be applied to **entire traces**.
+
+## High-Level Problem Statement
+
+### Ingest
+
+Implementing Dynamic Sampling comes with challenges, especially on the ingestion side of things.
+For Dynamic Sampling, we want to make sampling decisions for entire traces.
+However, to keep ingestion speedy, Relay only looks at singular transactions in isolation (as opposed to looking at whole traces).
+This means that we need the exact same decision basis for all transactions belonging to a trace.
+In other words, all transactions of a trace need to hold all of the information to make a sampling decision, and that information needs to be the same across all transactions of the trace.
+We call the information we base sampling decisions on **"Dynamic Sampling Context"** or **"DSC"**.
+
+### SDKs
+
+SDKs are responsible for propagating **Dynamic Sampling Context** across all applications that are part of a trace.
+This involves:
+
+1. Collecting the information that makes up the DSC **xor** extracting the DSC from incoming requests.
+2. Propagating DSC to downstream SDKs.
+3. Sending the DSC to Sentry via an envelope.
+
+Because there are quite a few things to keep in mind for DSC propagation and to avoid every SDK running into the same problems, we defined a [unified propagation mechanism](#unified-propagation-mechanism) (step-by-step instructions) that all SDK implementations should be able to follow.
+
+## Baggage
+
+We chose `baggage` as the propagation mechanism for DSC. ([w3c baggage spec](https://www.w3.org/TR/baggage/))
+Baggage is a standard HTTP header with URI encoded key-value pairs.
+
+For the propagation of DSC, SDKs first read the DSC from the baggage header of incoming requests/messages.
+To propagate DSC to downstream SDKs/services, we create a baggage header (or modify an existing one) through HTTP request instrumentation.
+
+<Alert level="info">
+
+Other vendors might also be using the `baggage` header.
+If a `baggage` header already exists on an outgoing request, SDKs should aim to be good citizens by only **appending** Sentry values to the header.
```

Comment:
For incoming `baggage` that already contains `sentry-` list-items:

* We do not modify existing ones as we want to keep them the way they were sent by the first SDK in line, right?
* Does that also mean we do not add missing list-items? Assuming `userid` was not set by the first SDK but another SDK knows it, should it add the list-item to `baggage` and other places or keep them the way they were sent by the first SDK?

Reply:
As Dynamic Sampling Context is immutable across an entire trace, we cannot add additional `sentry-` list items when there are incoming ones. The reasoning for immutability is explained in the #ingest section of this doc - essentially relay needs to make exactly the same sampling decision for all individual transactions of a trace. This requires all transactions to have the exact same Dynamic Sampling Context. 

I added the pseudo algorithm of how SDKs should instrument incoming and outgoing requests in regards to DSC. I hope that clarifies this a bit: 3f1302481d4cbbb0e364fdcf2fae34d8324ad297

Reply:
>We do not modify existing ones as we want to keep them the way they were sent by the first SDK in line, right?
Correct

> Does that also mean we do not add missing list-items?
Also Correct

Reply:
Also added some sentences to explain this a little bit more explicitly: ec30942000b971681c4767a8c4bed2aefa0f378e

Reply:
Let’s also make it very clear that we should not delete or change existing non-sentry baggage values (and this this very very important).

Reply:
> Does this mean that if an incoming baggage header is close to the limit of 8192 characters we are not allowed to add our sentry list-items to the header?

Yes we need to be good citizens and respect the standard here.

> Does that mean we should add them in an all or nothing fashion or do we have a priority for which of them we want to try and add until we run out of characters?

I don't think we have a priority, we try to add what we can. I'd stay away from the all-or-nothing approach since that seems it might be confusing to users (not sure though). I'm comfortable enough with this for now since the vast majority of head SDK's (those creating the head transactions of a trace) will be from browser/mobile, which should not have problems with incoming baggage - only outgoing. As a result, we won't really hit this to start, and when if it ever becomes a substantial problem, I think we can come back and revisit. For now, let's just optimize for getting this out the door.

Reply:
>I don't think we have a priority, we try to add what we can

I agree - let's try to keep things simple for now. In case this really becomes a problem, we could still revisit this and discuss priorities of keys or other handling strategies.

For now, if we exceed the max length, we should log a warning though so that users/we are aware of why DSC might be propagated/sent incomplete in this case. 
(Which is what we're [currently doing](https://github.com/getsentry/sentry-javascript/blob/06d573f3d68b21efea81aef222755f14b1c6805a/packages/utils/src/baggage.ts#L84-L90) in the JS SDKs).